### PR TITLE
fix: simp confluence issues with `Set.range`

### PR DESCRIPTION
**Pull request by timotree3**
`Set.mem_range` is a questionable simp lemma because it replaces `Set.range` with an existential unconditionally. This means that if any simp lemma applies to `Set.range`, there must be a corresponding lemma for existentials. These lemmas are what I found missing after [loogling](https://loogle.lean-lang.org/?q=Set.range%2C+Eq).

---

Some of the statements of the confluence "merge" lemmas are pretty ugly. e.g. `(Sum.inl y = x → ∃ y', Sum.inl y' = x) ↔ True`. If we agree simp confluence is important, maybe there's a better alternative. I wonder if `Set.mem_range` should be an aesop rule but not in the default simp set.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Option/Basic.lean [hunk 1]
```diff
@@ -426,4 +426,19 @@ theorem elim_comp₂ (h : α → β → γ) {f : γ → α} {x : α} {g : γ →
 theorem elim_apply {f : γ → α → β} {x : α → β} {i : Option γ} {y : α} :
     i.elim x f y = i.elim (x y) fun j => f j y := by rw [elim_comp fun f : α → β => f y]
 
+@[simp]
+theorem eq_none_or_exists_some_eq : ∀ x : Option α, x = none ∨ ∃ y, some y = x
```
Comment (timotree3) [inline Mathlib/Data/Option/Basic.lean:430]: If I remember correctly, this lemma already exists for `∃ y, x = some y`, which is some small evidence in favor of #9968

**Comment by timotree3** [comment]
Blocked on [simpNF issue](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/simpNF.20says.20lemma.20doesn't.20apply.20to.20own.20LHS/near/421746495)

**Comment by timotree3** [comment]
The more I think about this PR the more I feel like this is a horrible solution to the problem. These new simp lemmas are often equally questionable compared to `Set.mem_range`, and I haven't looked closely at each of them, but I'm guessing they create new confluence issues of their own. I think [the tests in this PR](https://github.com/leanprover-community/mathlib4/pull/10528/files#diff-9858d31e3195027d979e42885d393a1b906ac3d409ebb5e7aeb2202ecd1d9f12) should serve more as a demonstration of the issues with having `Set.mem_range` in the default simp-set, and I should just start a discussion on Zulip about how to actually fix them.

**Comment by grunweg** [comment]
Sounds like `awaiting-author` better reflects this PR's status.
awaiting-author

Feel free to apply other labels (e.g. `please-adopt`/`help-wanted`) if they fit!

**Comment by timotree3** [comment]
I don't like the simp lemmas this PR would add. Closing for now

**Pull request closed without merging**